### PR TITLE
test whether resuming from a checkpoint changes algorithm effect

### PR DESCRIPTION
**Pull request by growlix**
This PR introduces `tests/algorithms/test_algorithm_resumption.py`, which tests whether resuming from a checkpoint changes the effect of an algorithm compared to training without checkpointing and resumption. Currently fails on `layer_freezing`, `sam`, and `swa`.

### CODE REVIEW

File: tests/algorithms/test_algorithm_resumption.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        [42, "1ep", "ep{epoch}", "ep3", "ep5"],  # test save at epoch end
+        [42, "1ep", "ep{epoch}.tgz", "ep3.tgz", "ep5.tgz"],  # test tarball with compression
+        [42, "2ba", "ba{batch}", "ba12", "ba25"],  # test save batch in partial epoch
+        [42, "1ba", "ba{batch}", "ba15", "ba25"],  # test save final batch in epoch
+        [42, "2ba", "ba{batch}", "ba16", "ba25"],  # test save batch after complete epoch
+    ],
+)
+@pytest.mark.parametrize("model_name", [None, "resnet50_synthetic", "gpt2_52m"])
+@pytest.mark.parametrize("algorithm", algorithms_registry.items(), ids=algorithms_registry.keys())
+def test_algorithm_resumption(
```
Comment (hanlint) [inline tests/algorithms/test_algorithm_resumption.py:98]: Looking good! For readability, can you refactor this to not use the TrainerHparams object? e.g.
```
config = {
    'model': ...,
    'algorithms': ...,
}

trainer = Trainer(**config)
```

File: tests/algorithms/test_algorithm_resumption.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    [
+        [None, "1ep", "ep{epoch}", "ep3", "latest-rank{rank}"],  # test randomized seed saving and symlinking
+        [42, "1ep", "ep{epoch}", "ep3", "ep5"],  # test save at epoch end
+        [42, "1ep", "ep{epoch}.tgz", "ep3.tgz", "ep5.tgz"],  # test tarball with compression
+        [42, "2ba", "ba{batch}", "ba12", "ba25"],  # test save batch in partial epoch
+        [42, "1ba", "ba{batch}", "ba15", "ba25"],  # test save final batch in epoch
+        [42, "2ba", "ba{batch}", "ba16", "ba25"],  # test save batch after complete epoch
+    ],
+)
+@pytest.mark.parametrize("model_name", [None, "resnet50_synthetic", "gpt2_52m"])
```
Comment (hanlint) [inline tests/algorithms/test_algorithm_resumption.py:96]: `model_name=None` is a bit unclear, perhaps do `default` instead? The logic below should still work.

File: tests/algorithms/test_algorithm_resumption.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+    if deepspeed_enabled:
+        if not _is_archive(resume_file):
+            resume_file += ".tar"
+        if not _is_archive(final_checkpoint):
+            final_checkpoint += ".tar"
+
+    composer_trainer_hparams.schedulers = []
+    composer_trainer_hparams.max_duration = '5ep'
+    if model_name in ["resnet50_synthetic", "gpt2_52m"]:
```
Comment (hanlint) [inline tests/algorithms/test_algorithm_resumption.py:132]: should guard against unknown model_names here to reduce future bug potential (e.g. if someone puts in`unet`, it'll actually just run the default and all tests would pass)

File: tests/algorithms/test_algorithm_resumption.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        if not isinstance(device_hparams, GPUDeviceHparams):
+            pytest.skip("Real models require a GPU -- otherwise they take too long")
+        model_hparams = TrainerHparams.load(model_name)
+        composer_trainer_hparams.train_dataset = model_hparams.train_dataset
+        composer_trainer_hparams.val_dataset = model_hparams.val_dataset
+        composer_trainer_hparams.model = model_hparams.model
+        composer_trainer_hparams.optimizer = model_hparams.optimizer
+        composer_trainer_hparams.schedulers = model_hparams.schedulers
+    if not isinstance(composer_trainer_hparams.train_dataset, SyntheticHparamsMixin):
```
Comment (hanlint) [inline tests/algorithms/test_algorithm_resumption.py:141]: does this mean that even though `gpt2_52m` is in the model_name, it will always be skipped? if so, we may just want to remove gpt2_52m form the test suite.

File: tests/algorithms/test_algorithm_resumption.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    },
+    "cutmix": {
+        "num_classes": dummy_num_classes
+    },
+    "mixup": {
+        "num_classes": dummy_num_classes
+    }    
+}
+
+simple_net_skiplist = {
```
Comment (hanlint) [inline tests/algorithms/test_algorithm_resumption.py:36]: perhaps just have one skip list:
```
skip_list = {
    'default': {'colout': ...}
    'resnet50_synthetic': {, ..},
}
```

Then below in your code you can do this once:
```
if algo in skip_list[model_name]:
```
```

File: tests/algorithms/test_algorithm_resumption.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        if not _is_archive(final_checkpoint):
+            final_checkpoint += ".tar"
+
+    composer_trainer_hparams.schedulers = []
+    composer_trainer_hparams.max_duration = '5ep'
+    if model_name in ["resnet50_synthetic", "gpt2_52m"]:
+        if not isinstance(device_hparams, GPUDeviceHparams):
+            pytest.skip("Real models require a GPU -- otherwise they take too long")
+        model_hparams = TrainerHparams.load(model_name)
+        composer_trainer_hparams.train_dataset = model_hparams.train_dataset
```
Comment (hanlint) [inline tests/algorithms/test_algorithm_resumption.py:136]: ah i can see why it's easier to use the trainer hparams objects now (please disregard my previous comment). A few suggestions:
* since we basically overwrite all the defaults from the `composer_trainer_hparams` object, can you just create it directly here? e.g.
```
hparams = TrainerHparams(...)
```
this change would help readability without having to refer to the fixture file.

* use a much shorter variable name (e.g. `trainer_hparams`, or just `hparams`).

File: tests/algorithms/test_algorithm_resumption.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    """strategy:
+    - train five epochs. capture checkpoints after `checkpoint_interval` and ep5.
+    - create a new trainer from the `checkpoint_interval` checkpoint, and train until end. checkpoint again.
+    - assert that the checkpoint from the new trainer at the end is the same as the checkpoint from the first trainer at the end.
+    """
+    del world_size  # unused. Read via env variable
+
+    algo_name, algo_hparams = algorithm
+
+    if not isinstance(device_hparams, GPUDeviceHparams) and deepspeed_enabled:
```
Comment (hanlint) [inline tests/algorithms/test_algorithm_resumption.py:121]: Is this check needed, given that the mark.parameterize above only does deepspeed in combination with GPU?

File: tests/algorithms/test_algorithm_resumption.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        return
+
+    if model_name is None and algo_name in simple_net_skiplist:
+        pytest.skip(simple_net_skiplist[algo_name])
+
+    if model_name == 'resnet50_synthetic' and algo_name in resnet_skiplist:
+        pytest.skip(resnet_skiplist[algo_name])
+    
+    if model_name == 'gpt2_52m' and algo_name in lm_skiplist:
+        pytest.skip(lm_skiplist[algo_name])
```
Comment (hanlint) [inline tests/algorithms/test_algorithm_resumption.py:155]: These can be condensed into:
```
if algo_name in skip_list[model_name]:
    pytest.skip(skip_list[model_name][algo_name])
```

File: tests/algorithms/test_algorithm_resumption.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+    composer_trainer_hparams.validate_every_n_batches = 1 if resume_file.startswith("ba") else 0
+    composer_trainer_hparams.validate_every_n_epochs = 1 if resume_file.startswith("ep") else 0
+    first_trainer = composer_trainer_hparams.initialize_object()
+    first_trainer.fit()
+    save_interval_time = Time.from_timestring(save_interval)
+    if save_interval_time.unit == TimeUnit.EPOCH:
+        expected_num_checkpoints = ((num_epochs - 1) // save_interval_time.value) + 1
+    else:
+        expected_num_checkpoints = (
+            (composer_trainer_hparams.train_subset_num_batches * num_epochs - 1) // save_interval_time.value) + 1
```
Comment (hanlint) [inline tests/algorithms/test_algorithm_resumption.py:205]: Since this path is tested elsewhere, let's remove this code and the asserts below. The unit tests should just test one thing -- algorithm resumption.

**Review by hanlint — COMMENTED** [review]
Looks good, mostly comments to simplify the test and code for readability.

File: tests/algorithms/test_algorithm_resumption.py (hunk 3 above)
Comment (growlix) [reply tests/algorithms/test_algorithm_resumption.py:132]: Changed to `if model_name != "default"`

File: tests/algorithms/test_algorithm_resumption.py (hunk 4 above)
Comment (growlix) [reply tests/algorithms/test_algorithm_resumption.py:141]: Are `SyntheticHparamsMixin` exclusive to vision datasets?

File: tests/algorithms/test_algorithm_resumption.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+    checkpoint_saver = None
+    for callback in first_trainer.state.callbacks:
+        if isinstance(callback, CheckpointSaver):
+            checkpoint_saver = callback
+    assert checkpoint_saver is not None
+    assert len(checkpoint_saver.saved_checkpoints) == expected_num_checkpoints
+    checkpoint_a_file_path = os.path.join(checkpoint_a_folder, resume_file)
+    checkpoint_b_file_path = os.path.join(run_directory.get_node_run_directory(), "rank_{rank}", checkpoint_a_folder,
+                                          final_checkpoint)
```
[the remark this answers is not in the document]
Comment (growlix) [reply tests/algorithms/test_algorithm_resumption.py:214]: The rank is part of the file path, but it is always rank 0 so could change it to `"rank_0"`.

File: tests/algorithms/test_algorithm_resumption.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    second_trainer_hparams.save_folder = checkpoint_b_folder
+    second_trainer_filepath = os.path.join(run_directory.get_node_run_directory(), "rank_{rank}",
+                                           checkpoint_a_file_path)
+    second_trainer_hparams.load_path_format = second_trainer_filepath
+    second_trainer_hparams.load_weights_only = False
+    second_trainer_hparams.load_strict_model_weights = False
+
+    second_trainer = second_trainer_hparams.initialize_object()
+    second_trainer.fit()
+    checkpoint_c_file_path = os.path.join(run_directory.get_node_run_directory(), "rank_{rank}", checkpoint_b_folder,
```
[the remark this answers is not in the document]
Comment (growlix) [reply tests/algorithms/test_algorithm_resumption.py:228]: I could be mistaken, but I thought this creates a second directory so that the second (resumed) training run doesn't overwrite the checkpoints from the first run.

File: tests/fixtures/models.py [hunk 12]
```diff
@@ -35,9 +35,9 @@ def __init__(self, num_channels: int, num_classes: int) -> None:
         # These tests attempt to perform surgery on `fc1` layer, and we want
         # to make sure that post-surgery, self.fc1 refers to the same parameters
         # as self.net[1]
-        self.fc1 = torch.nn.Linear(num_channels, 5)
+        self.fc1 = torch.nn.Linear(num_channels, 20)
```
Comment (hanlint) [inline tests/fixtures/models.py:38]: this may slow down other tests -- is this strictly neccessary?

**Review by hanlint — APPROVED** [review]
LGTM

File: tests/fixtures/models.py (hunk 12 above)
Comment (growlix) [reply tests/fixtures/models.py]: Switched back

File: tests/fixtures/new_fixtures.py [hunk 13]
```diff
@@ -33,3 +35,12 @@ def empty_logger(minimal_state: State) -> Logger:
...
     monkeypatch.setenv("WANDB_MODE", "disabled")
+
+
+# Class-scoped temporary directory. That deletes itself. This is useful for e.g. not
+# writing too many checkpoints.
+@pytest.fixture(scope='class')
+def self_destructing_tmp(tmp_path_factory: pytest.TempPathFactory):
+    my_tmpdir = tmp_path_factory.mktemp("checkpoints")
+    yield my_tmpdir
+    shutil.rmtree(str(my_tmpdir))
```
Comment (ravi-mosaicml) [inline tests/fixtures/new_fixtures.py:46]: If the test fails, do we lose the tempdir? (It is helpful to keep it around for debugging a failing test)

File: tests/fixtures/dummy_fixtures.py [hunk 14]
```diff
@@ -42,6 +42,11 @@ def dummy_val_batch_size() -> int:
     return 32
 
 
+@pytest.fixture()
+def dummy_train_n_samples() -> int:
+    return 1000
```
Comment (ravi-mosaicml) [inline tests/fixtures/dummy_fixtures.py:47]: Curious if anything is using this fixture?

File: tests/algorithms/test_algorithm_resumption.py [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+        ],  # test tarball with compression
+        [42, "2ba", "ba{batch}-rank{rank}", "ba12-rank{rank}", "ba25-rank{rank}"],  # test save batch in partial epoch
+        [42, "1ba", "ba{batch}-rank{rank}", "ba15-rank{rank}", "ba25-rank{rank}"],  # test save final batch in epoch
+        [42, "2ba", "ba{batch}-rank{rank}", "ba16-rank{rank}", "ba25-rank{rank}"
+        ],  # test save batch after complete epoch
+    ],
+)
+@pytest.mark.parametrize("model_name", ["default", "resnet50_synthetic", "gpt2_52m"])
+@pytest.mark.parametrize("algorithm", algorithms_registry.items(), ids=algorithms_registry.keys())
+def test_algorithm_resumption(device_hparams: DeviceHparams, world_size: int, deepspeed_enabled: bool,
```
Comment (ravi-mosaicml) [inline tests/algorithms/test_algorithm_resumption.py:96]: This test case seems pretty similar to the existing checkpointing test. Would it be possible to refactor the tests to share logic where possible?

File: tests/algorithms/test_algorithm_resumption.py [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+
+@pytest.mark.timeout(180)
+@pytest.mark.parametrize("world_size", [
+    pytest.param(1),
+    pytest.param(2, marks=pytest.mark.world_size(2)),
+])
+@pytest.mark.parametrize("device_hparams,deepspeed_enabled,zero_stage", [
+    pytest.param(CPUDeviceHparams(), False, None, id="cpu-ddp"),
+    pytest.param(GPUDeviceHparams(), False, None, id="gpu-ddp", marks=pytest.mark.gpu),
+    pytest.param(GPUDeviceHparams(), True, 0, id="deepspeed-zero0", marks=pytest.mark.gpu),
```
Comment (ravi-mosaicml) [inline tests/algorithms/test_algorithm_resumption.py:93]: Is it necessary to check serialization on deepspeed? I don't think deepspeed would change anything w.r.t. an individual algorithm (the standard DDP tests ensure that algorithms do not assume that the model is a ComposerModel)

File: tests/algorithms/test_algorithm_resumption.py [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+)
+# @pytest.mark.parametrize("model_name", ["default", "resnet50_synthetic", "gpt2_52m"])
+@pytest.mark.parametrize("model_name", ["resnet50_synthetic"])
+@pytest.mark.parametrize("algorithm", algorithms_registry.items(), ids=algorithms_registry.keys())
+def test_algorithm_resumption(device_hparams: DeviceHparams, world_size: int, deepspeed_enabled: bool,
+                              zero_stage: Optional[int], save_interval: str, save_filename: str, resume_file: str,
+                              final_checkpoint: str, seed: Optional[int], model_name: str,
+                              algorithm: Tuple[str, Type[AlgorithmHparams]], dummy_model_hparams: ModelHparams,
+                              dummy_train_dataset_hparams: DatasetHparams, dummy_val_dataset_hparams: DatasetHparams,
+                              tmpdir: py.path.local):
```
Comment (ravi-mosaicml) [inline tests/algorithms/test_algorithm_resumption.py:119]: Use the new fixture :)

```suggestion
                              self_destructing_tmp: py.path.local):
```

Can you also update `test_checkpoint` to use this fixture?

File: tests/algorithms/test_algorithm_resumption.py [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+    pytest.param(GPUDeviceHparams(), True, 1, id="deepspeed-zero1", marks=pytest.mark.gpu),
+    pytest.param(GPUDeviceHparams(), True, 2, id="deepspeed-zero2", marks=pytest.mark.gpu),
+])
+@pytest.mark.parametrize(
+    "seed,save_interval,save_filename,resume_file,final_checkpoint",
+    [
+        [None, "1ep", "ep{epoch}-rank{rank}", "ep2-rank{rank}", "latest-rank{rank}"
+        ],  # test randomized seed saving and symlinking
+        [42, "1ep", "ep{epoch}-rank{rank}", "ep2-rank{rank}", "ep4-rank{rank}"],  # test save at epoch end
+        # [42, "1ep", "ep{epoch}-rank{rank}.tgz", "ep2-rank{rank}.tgz", "ep4-rank{rank}.tgz"
```
Comment (ravi-mosaicml) [inline tests/algorithms/test_algorithm_resumption.py:103]: Before merging, can you remove the commented code?

**Review by ravi-mosaicml — APPROVED** [review]
LGTM; see comments.
